webapp/src/tutorialValidator.ts

```diff
@@ -0,0 +1,112 @@
+/// <reference path="../../built/pxtlib.d.ts" />
+
+import * as compiler from "./compiler";
+
+export enum TutorialCodeStatus {
+    Unknown = "Unknown",
+    Invalid = "Invalid",
+    Valid = "Valid"
+}
+
+/**
+* Check the user's code to the tutorial and returns a Tutorial status of the user's code
+* @param step the current tutorial step
+* @param source XML of workspace
+* @param workspaceBlocks Blockly blocks used of workspace
+* @param blockinfo Typescripts of the workspace
+* @return A TutorialCodeStatus
+*/
+export async function validate(step: pxt.tutorial.TutorialStepInfo, source: string, workspaceBlocks: Blockly.Block[], blockinfo: pxtc.BlocksInfo): Promise<TutorialCodeStatus> {
+    // Check to make sure blocks are in the workspace
+    if (workspaceBlocks.length > 0) {
+        let userBlockTypes: string[] = workspaceBlocks.map(b => b.type);
+        let tutorialBlockTypes: string[] = await tutorialBlockList(step, blockinfo);
+        let usersBlockUsed: pxt.Map<number> = blockCount(userBlockTypes);
+        let tutorialBlockUsed: pxt.Map<number> = blockCount(tutorialBlockTypes);
+        if (!validateNumberOfBlocks(usersBlockUsed, tutorialBlockUsed)) {
+            return TutorialCodeStatus.Invalid;
+        }
+        return TutorialCodeStatus.Valid;
+    }
+    return TutorialCodeStatus.Unknown;
+}
+
+/**
+* Loops through an array of blocks and returns a map of blocks and the count for that block
+* @param arr a string array of blocks
+* @return a map <Block type, frequency>
+*/
+function blockCount(arr: string[]): pxt.Map<number> {
+    let frequencyMap: pxt.Map<number> = {};
+    for (let i: number = 0; i < arr.length; i++) {
+        let blockIndex = hasBlock(arr, i, arr[i]);
+        if (blockIndex == i) {
+            frequencyMap[arr[i]] = 0;
+        }
+        frequencyMap[arr[i]] = frequencyMap[arr[i]] + 1;
+    }
+    return frequencyMap;
+}
+
+/**
+* Checks to see if the array has the target value and returns the target value's index
+* @param arr an array of strings
+* @param index the next available index
+* @param val the target value
+* @return the index 
+*/
+function hasBlock(arr: string[], index: number, val: string): number {
+    for (let i: number = 0; i < index; i++) {
+        if (arr[i] === val) {
+            return i;
+        }
+    }
+    return index;
+}
+
+/**
+* Checks to see if the array has the target value and returns the target value's index
+* @param step the current tutorial step
+* @param blockinfo Typescripts of the workspace
+* @return the tutorial blocks
+*/
+function tutorialBlockList(step: pxt.tutorial.TutorialStepInfo, blocksInfo: pxtc.BlocksInfo): Promise<string[]> {
+    let hintContent = step.hintContentMd;
+    hintContent = hintContent.replace("\`\`\`blocks", "");
```

Comment:
Does this function behave nicely if a step doesn't have a snippet as a hint? (e.g. it may have a gif instead?) for example step 15 in https://arcade.makecode.com/#tutorial:/tutorials/catch-the-football (or localhost:3232/#tutorial:/tutorials/catch-the-football)